File: 02-Python全栈AI刷题学习小程序/backend/app/services/quiz_service.py
from datetime import date

from fastapi import HTTPException

from app.schemas import (
    AnswerResult,
    ChallengeHistoryItem,
    DailyChallenge,
    LearningProfile,
    Quiz,
    Report,
    ReportHistoryItem,
    WrongQuestion,
)
from app.services.ai_service import generate_quiz
from app.storage.memory_store import store


def create_quiz(session_id: str, content: str) -> Quiz:
    if not content.strip():
        raise HTTPException(status_code=422, detail="content is required")

    quiz = generate_quiz(content)
    store.save_quiz(session_id, quiz)
    return quiz


def submit_answer(
    session_id: str,
    quiz_id: str,
    question_id: str,
    answer: list[str],
) -> AnswerResult:
    quiz = store.get_quiz(quiz_id)
    if quiz is None:
        raise HTTPException(status_code=404, detail="quiz not found")

    question = next((item for item in quiz.questions if item.id == question_id), None)
    if question is None:
        raise HTTPException(status_code=404, detail="question not found")

    normalized_answer = sorted(answer)
    correct_answer = sorted(question.answer)
    result = AnswerResult(
        questionId=question.id,
        isCorrect=normalized_answer == correct_answer,
        correctAnswer=question.answer,
        explanation=question.explanation,
        knowledge_point=question.knowledge_point,
    )
    store.save_answer(session_id, quiz_id, result)
    store.record_answer_stat(session_id, result.isCorrect)
    if result.isCorrect:
        store.remove_wrong_question(session_id, quiz_id, question.id)
    else:
        store.save_wrong_question(
            session_id,
            WrongQuestion(
                quizId=quiz_id,
                questionId=question.id,
                stem=question.stem,
                selectedAnswer=answer,
                correctAnswer=question.answer,
                explanation=question.explanation,
                knowledge_point=question.knowledge_point,
            ),
        )
    return result


def generate_report(session_id: str, quiz_id: str) -> Report:
    quiz = store.get_quiz(quiz_id)
    if quiz is None:
        raise HTTPException(status_code=404, detail="quiz not found")

    answers = store.get_answers(session_id, quiz_id)
    score = sum(1 for answer in answers if answer.isCorrect)
    total = len(quiz.questions)
    mastery = round(score / total * 100) if total else 0
    weak_points = [
        answer.knowledge_point for answer in answers if not answer.isCorrect
    ] or ["暂无明显薄弱点"]

    report = Report(
        quizId=quiz.quizId,
        title=quiz.title,
        score=score,
        total=total,
        mastery=mastery,
        summary=f"本轮围绕“{quiz.title}”完成 {total} 道题，答对 {score} 道。",
        weakPoints=weak_points,
        nextSteps=[
            "复习错题对应的知识点。",
            "用自己的话重新解释本轮主题。",
            "再生成一组题巩固薄弱点。",
        ],
    )
    store.save_report(session_id, report)
    return report


def get_wrong_questions(session_id: str) -> list[WrongQuestion]:
    return store.get_wrong_questions(session_id)


def get_report_history(session_id: str) -> list[ReportHistoryItem]:
    return [
        ReportHistoryItem(
            quizId=report.quizId,
            title=report.title,
            score=report.score,
            total=report.total,
            mastery=report.mastery,
            weakPoints=report.weakPoints,
        )
        for report in store.get_reports(session_id)
    ]


def get_challenge_history(session_id: str) -> list[ChallengeHistoryItem]:
    return [
        ChallengeHistoryItem(
            quizId=report.quizId,
            title=report.title,
            score=report.score,
            total=report.total,
            mastery=report.mastery,
            completedAt=store.get_report_time(session_id, report.quizId),
        )
        for report in store.get_reports(session_id)
    ]


def get_daily_challenge(session_id: str) -> DailyChallenge:
    today = date.today().isoformat()
    reports = store.get_reports(session_id)
    today_quiz_ids = {report.quizId for report in reports}
    answered = sum(
        len(store.get_answers(session_id, quiz_id)) for quiz_id in today_quiz_ids
    )
    correct = sum(
        sum(1 for answer in store.get_answers(session_id, quiz_id) if answer.isCorrect)
        for quiz_id in today_quiz_ids
    )
    target = 5
    capped_answered = min(answered, target)
    return DailyChallenge(
        date=today,
        target=target,
        answered=capped_answered,
        correct=correct,
        completed=answered >= target,
        progress=round(capped_answered / target * 100),
    )


def get_learning_profile(session_id: str) -> LearningProfile:
    return store.get_profile(session_id)


def regenerate_weak_quiz(session_id: str, quiz_id: str) -> Quiz:
    quiz = store.get_quiz(quiz_id)
    if quiz is None:
        raise HTTPException(status_code=404, detail="quiz not found")

    wrong_questions = [
        item for item in store.get_wrong_questions(session_id) if item.quizId == quiz_id
    ]
    weak_points = [item.knowledge_point for item in wrong_questions]
    if not weak_points:
        weak_points = [
            answer.knowledge_point
            for answer in store.get_answers(session_id, quiz_id)
        ]
    weak_text = "、".join(dict.fromkeys(weak_points)) or quiz.title
    content = (
        f"请围绕这些薄弱点重新生成一组复习题：{weak_text}。"
        "题目要适合初学者，并重点检查用户是否真的理解。"
    )
    next_quiz = generate_quiz(content)
    store.save_quiz(session_id, next_quiz)
    return next_quiz

File: 02-Python全栈AI刷题学习小程序/frontend/src/services/quizService.ts
import Taro from '@tarojs/taro'

const API_BASE_URLS = ['http://127.0.0.1:8010', 'http://localhost:8010']

export type QuestionType = 'single' | 'multiple' | 'judge'
export type Difficulty = 'easy' | 'medium' | 'hard'

export interface Option {
  key: string
  text: string
}

export interface Question {
  id: string
  type: QuestionType
  stem: string
  options: Option[]
  answer: string[]
  explanation: string
  knowledge_point: string
  difficulty: Difficulty
}

export interface Quiz {
  quizId: string
  title: string
  summary: string
  questions: Question[]
}

export interface AnswerResult {
  questionId: string
  isCorrect: boolean
  correctAnswer: string[]
  explanation: string
  knowledge_point: string
}

export interface Report {
  quizId: string
  title: string
  score: number
  total: number
  mastery: number
  summary: string
  weakPoints: string[]
  nextSteps: string[]
}

export interface WrongQuestion {
  quizId: string
  questionId: string
  stem: string
  selectedAnswer: string[]
  correctAnswer: string[]
  explanation: string
  knowledge_point: string
  reviewed: boolean
}

export interface ReportHistoryItem {
  quizId: string
  title: string
  score: number
  total: number
  mastery: number
  weakPoints: string[]
}

export interface ChallengeHistoryItem {
  quizId: string
  title: string
  score: number
  total: number
  mastery: number
  completedAt: string
}

export interface DailyChallenge {
  date: string
  target: number
  answered: number
  correct: number
  completed: boolean
  progress: number
}

export interface Badge {
  id: string
  name: string
  description: string
  unlocked: boolean
}

export interface LearningProfile {
  level: number
  exp: number
  streakDays: number
  totalAnswered: number
  totalCorrect: number
  badges: Badge[]
}

interface ApiResponse<T> {
  code: number
  message: string
  data: T
}

export async function request<T>(
  url: string,
  method: 'GET' | 'POST',
  data?: unknown,
  extraHeader: Record<string, string> = {}
): Promise<T> {
  let lastError: unknown
  const token = Taro.getStorageSync<string>('authToken')
  const authHeader = token ? { Authorization: `Bearer ${token}` } : {}

  for (const baseUrl of API_BASE_URLS) {
    try {
      const response = await Taro.request<ApiResponse<T>>({
        url: `${baseUrl}${url}`,
        method,
        data,
        header: {
          'Content-Type': 'application/json',
          ...authHeader,
          ...extraHeader
        }
      })

      if (response.statusCode >= 400 || response.data.code !== 0) {
        throw new Error(response.data?.message || `接口返回异常：${response.statusCode}`)
      }

      return response.data.data
    } catch (error) {
      lastError = error
    }
  }

  const detail = lastError instanceof Error ? lastError.message : '未知网络错误'
  throw new Error(`后端接口暂时无法访问，请确认本地后端已启动，并在微信开发者工具勾选“不校验合法域名”。${detail}`)
}

export function getOrCreateSessionId(): string {
  const existing = Taro.getStorageSync<string>('sessionId')
  if (existing) return existing

  const sessionId = `session-${Date.now()}-${Math.random().toString(16).slice(2)}`
  Taro.setStorageSync('sessionId', sessionId)
  return sessionId
}

export function generateQuiz(content: string): Promise<Quiz> {
  return request<Quiz>('/api/generate-quiz', 'POST', {
    sessionId: getOrCreateSessionId(),
    inputType: 'text',
    content,
    questionCount: 5
  })
}

export function submitAnswer(quizId: string, questionId: string, answer: string[]): Promise<AnswerResult> {
  return request<AnswerResult>('/api/submit-answer', 'POST', {
    sessionId: getOrCreateSessionId(),
    quizId,
    questionId,
    answer
  })
}

export function generateReport(quizId: string): Promise<Report> {
  return request<Report>('/api/generate-report', 'POST', {
    sessionId: getOrCreateSessionId(),
    quizId
  })
}

export function getWrongQuestions(): Promise<WrongQuestion[]> {
  return request<WrongQuestion[]>(`/api/wrong-questions?sessionId=${getOrCreateSessionId()}`, 'GET')
}

export function regenerateWeakQuiz(quizId: string): Promise<Quiz> {
  return request<Quiz>('/api/regenerate-weak-quiz', 'POST', {
    sessionId: getOrCreateSessionId(),
    quizId
  })
}

export function getReportHistory(): Promise<ReportHistoryItem[]> {
  return request<ReportHistoryItem[]>(`/api/report-history?sessionId=${getOrCreateSessionId()}`, 'GET')
}

export function getChallengeHistory(): Promise<ChallengeHistoryItem[]> {
  return request<ChallengeHistoryItem[]>(`/api/challenge-history?sessionId=${getOrCreateSessionId()}`, 'GET')
}

export function getDailyChallenge(): Promise<DailyChallenge> {
  return request<DailyChallenge>(`/api/daily-challenge?sessionId=${getOrCreateSessionId()}`, 'GET')
}

export function getLearningProfile(): Promise<LearningProfile> {
  return request<LearningProfile>(`/api/profile?sessionId=${getOrCreateSessionId()}`, 'GET')
}
